src/java.base/share/classes/java/util/zip/ZipFile.java

```diff
@@ -1816,14 +1825,19 @@ private void initCEN(int knownTotal) throws IOException {
                     signatureMetaNames[j] = signatureNames.get(j);
                 }
             }
-            if (metaVersionsSet != null) {
-                metaVersions = new int[metaVersionsSet.size()];
-                int c = 0;
-                for (Integer version : metaVersionsSet) {
-                    metaVersions[c++] = version;
+            if (metaVersionsMap != null) {
+                metaVersions = new HashMap<>();
+                for (var entry : metaVersionsMap.entrySet()) {
+                    // Convert TreeSet<Integer> to int[] for performance
```

Comment:
I think if you just require order upon final freezing, you can just use a HashSet or ArrayList, and perform the sorting and deduplication when you compile to an int array.

Reply:
I like simple. 

Mixing Set.of and HashSet means a lot of semantic fu-fu (throwing/not throwing on nulls) and risks being suboptimal due making some call sites megamorphic.

Reply:
Yep, looks good. And I agree with Claes' evaluation that we should not complicate this construction further, especially that the result is quickly garbage collected when we compile into the string to int array map.

Reply:
We could also just use `Map<String, int[]>`. This would allow us to skip the whole `Set<Integer>` to `int[]`  transition step, and the temporary `HashMap` is no longer needed.

Deduplication could occur during registration:

```java
metaVersions.merge(name, new int[] {version}, (current, val) -> {
    // Check duplicates
    for (int v : current) {
        if (v == version) {
            return current;
        }
    }
    // Add version
    int[] merged = Arrays.copyOf(current, current.length + 1);
    merged[merged.length - 1] = version;
    return merged;
});
```

while the postprocessing step would simply sort each array:

```java
for (int[] versions : metaVersions.values()) {
    // JarFile::getVersionedEntry expects sorted versions
    Arrays.sort(versions);
}
```

This performs ~10% better on a version-heavy `ZipFileOpen`, and I think the code is reasonably simple compared to the current state of the PR. Thoughts?

Reply:
I guess performance leans on there being only one or a small number of versions per entry. Which I think is a fair assumption. Still would be good to have a stress test with many entries having many versions and make sure we don't regress such cases too much, no?

Reply:
> I guess performance leans on there being only one or a small number of versions per entry. 

I checked Spring Petclinic:

* 109 JAR dependencies
* 15 are multi-release enabled.
* The 15 MR-JARs have 9347 entries, of which 21 are versioned
* 9 of 15 MR JARs use the feature exclusively to hide `module-info.class` from Java 8.
* 3 of the 6 remaining JAR files version a single class file
* The remaining JARs mostly version on the 9 release.
* The maximum number of versioned entries in a is 5.
* Zero entries have more than a single version.

> Which I think is a fair assumption. Still would be good to have a stress test with many entries having many versions and make sure we don't regress such cases too much, no?

Such a test would be somewhat speculative. If I add 5 versions of each entry in `ZipFileOpen`, then yes, this PR sees a regression over mainline:

Baseline:
```
Benchmark                     (size)  Mode  Cnt       Score       Error  Units
ZipFileOpen.openCloseZipFile    1024  avgt   15  718234.380 ? 29454.645  ns/op
```

PR:
```
Benchmark                     (size)  Mode  Cnt        Score       Error  Units
ZipFileOpen.openCloseZipFile    1024  avgt   15  1284944.601 ? 74919.522  ns/op
```

While `JarFileGetEntry` improves over baseline:

Baseline:
```
Benchmark                              (mr)  (size)  Mode  Cnt    Score   Error  Units
JarFileGetEntry.getEntryHit           false    1024  avgt   15   68.984 ? 2.791  ns/op
JarFileGetEntry.getEntryHit            true    1024  avgt   15  324.763 ? 9.690  ns/op
JarFileGetEntry.getEntryHitUncached   false    1024  avgt   15   95.884 ? 1.404  ns/op
JarFileGetEntry.getEntryHitUncached    true    1024  avgt   15  333.491 ? 3.576  ns/op
JarFileGetEntry.getEntryMiss          false    1024  avgt   15   37.358 ? 2.215  ns/op
JarFileGetEntry.getEntryMiss           true    1024  avgt   15  650.467 ? 5.116  ns/op
JarFileGetEntry.getEntryMissUncached  false    1024  avgt   15   61.059 ? 0.382  ns/op
JarFileGetEntry.getEntryMissUncached   true    1024  avgt   15  678.457 ? 4.390  ns/op
```

PR:

```
Benchmark                              (mr)  (size)  Mode  Cnt    Score    Error  Units
JarFileGetEntry.getEntryHit           false    1024  avgt   15   61.333 ?  3.665  ns/op
JarFileGetEntry.getEntryHit            true    1024  avgt   15  153.987 ?  8.889  ns/op
JarFileGetEntry.getEntryHitUncached   false    1024  avgt   15   90.008 ?  4.158  ns/op
JarFileGetEntry.getEntryHitUncached    true    1024  avgt   15  168.831 ?  1.676  ns/op
JarFileGetEntry.getEntryMiss          false    1024  avgt   15   34.525 ?  1.878  ns/op
JarFileGetEntry.getEntryMiss           true    1024  avgt   15   61.036 ?  2.568  ns/op
JarFileGetEntry.getEntryMissUncached  false    1024  avgt   15   75.246 ? 20.375  ns/op
JarFileGetEntry.getEntryMissUncached   true    1024  avgt   15   81.484 ?  1.981  ns/op
```

Future work could probably improve this situation by shifting version resolution to `initCEN` time. That way, only the single version relevant to the `JarFile` `featureVersion` would need to be considered. That's not so easy to do currently, since this happens in `JarFile`s super constructor. But with flexible constructor bodies, we may be able to do something. But not for this PR.

Reply:
I poked around with an alternative approach that uses `java.util.BitSet`. This also allows us to use the map we build up as-is, avoids sorting et.c. Performance on `JarFileGetEntry` (#21489 version) is the same in my testing, but performance on `ZipFileOpen` with your 5 versions of each entry stress test should be much improved (might beat the baseline). Can you share your local changes to that micro in a gist or e-mail so I can verify?

Reply:
With this change to `ZipFileOpen`: 
```
diff --git a/test/micro/org/openjdk/bench/java/util/zip/ZipFileOpen.java b/test/micro/org/openjdk/bench/java/util/zip/ZipFileOpen.java
index 4f6ae6373ec..0e22cf7f0ab 100644
--- a/test/micro/org/openjdk/bench/java/util/zip/ZipFileOpen.java
+++ b/test/micro/org/openjdk/bench/java/util/zip/ZipFileOpen.java
@@ -73,6 +73,11 @@ public void beforeRun() throws IOException {

                 ename += entryPrefixes[i % entryPrefixes.length] + "-" + i;
                 zos.putNextEntry(new ZipEntry(ename));
+                zos.putNextEntry(new ZipEntry("META-INF/versions/9/" + ename));
+                zos.putNextEntry(new ZipEntry("META-INF/versions/11/" + ename));
+                zos.putNextEntry(new ZipEntry("META-INF/versions/17/" + ename));
+                zos.putNextEntry(new ZipEntry("META-INF/versions/21/" + ename));
+                zos.putNextEntry(new ZipEntry("META-INF/versions/24/" + ename));
             }
         }
         zipFile = tempFile;
```

On that modified `ZipFileOpen` micro,
https://github.com/eirbjo/jdk/pull/1 improves over #21489 but still regress compared to the baseline:

```
Benchmark                       (size)  Mode  Cnt        Score       Error  Units
ZipFileOpen.openCloseZipFile      1024  avgt   15   677386,387 ± 15761,989  ns/op # baseline
ZipFileOpen.openCloseZipFile      1024  avgt   15  1121127,629 ± 16716,218  ns/op # pull/21489, 0.6x
ZipFileOpen.openCloseZipFile      1024  avgt   15   968501,630 ± 28761,067  ns/op # eirbjo/jdk#1, 0.7x
```

I suspect remaining regression is mostly from the added `String` creation/processing in `initCEN`.

Reply:
> I suspect remaining regression is mostly from the added `String` creation/processing in `initCEN`.

Yes, there is a cost with the String construction when parsing the name out of `META-INF/versions/{name}`

Reply:
There are a few possible strategies for avoiding that additional parse, since the effect we're getting at here is to have a quick filter to avoid pointless lookups and not necessarily an exact mapping.

One is to store the `checkedHash` rather than the full `String`. This gets `openCloseZipFile` down to ~910000 ns/op. (`checkedHash` very hot in profiles). There's still a chance for redundant lookups on hash collisions, but this should be rare.

Another is to store a `BitSet` per name length. This gets `ZipFileOpen` down to baseline level (~670000 ns/op), but increases chance of having to do redundant lookups a lot.

Both also improves footprint (not keeping each versioned entry `String` in memory would be nice).

Reply:
> One is to store the `checkedHash` rather than the full `String`. This gets `openCloseZipFile` down to ~910000 ns/op. (`checkedHash` very hot in profiles). There's still a chance for redundant lookups on hash collisions, but this should be rare.

Seems like a resonable trade-off.  Could you take a look at the latest 771488e and see if that represents your suggestion here?

Reply:
Yes, 771488e matches my quick experiment (sans some cleanups you've made). I agree this variant makes for a reasonable trade-off.

Reply:
Maybe there's some benefit in detecting "META-INF/versions/" in the `checkAndAddEntry` method and then calculate the two different hash values back to back. Hash calculations are fast on modern hardware with vector support, but the memory accesses might not be so we might get a general win from re-arranging some of this stuff. I think this is more suitable for a follow-up experiment, though.